src/v2i-hub/MessageLoggerPlugin/src/MessageLoggerPlugin.cpp

```diff
@@ -0,0 +1,366 @@
+	// This method does NOT execute in the main thread, so variables must be protected
+	// (e.g. using std::atomic, std::mutex, etc.).
+
+	int instance;
+	GetConfigValue("Instance", instance);
+
+	GetConfigValue("Frequency", __frequency_mon.get());
+	__frequency_mon.check();
+
+	std::lock_guard<mutex> lock(_cfgLock);
+	GetConfigValue("File Location", _fileDirectory);
+	GetConfigValue("File Size In MB", _maxFilesizeInMB);
+	GetConfigValue("Messagetype", _cvmsgtype);
+	GetConfigValue("Filename", _filename);
+	std::string oldFilename = _curFilename;
+	_curFilename = _fileDirectory + "/" + _filename + ".json";
+
+	if (_curFilename.compare (oldFilename) !=0 )
+	{
+		_logFile.close();
+		OpenMSGLogFile();
+
+	}
+
+}
+
+/**
+ * Called when configuration is changed
+ *
+ * @param key Key of the configuration value changed
+ * @param value Changed value
+ */
+void MessageLoggerPlugin::OnConfigChanged(const char *key, const char *value)
+{
+	PluginClient::OnConfigChanged(key, value);
+	UpdateConfigSettings();
+}
+
+/**
+ * Called on plugin state change
+ *
+ * @para state New plugin state
+ */
+void MessageLoggerPlugin::OnStateChange(IvpPluginState state)
+{
+	PluginClient::OnStateChange(state);
+
+	if (state == IvpPluginState_registered)
+	{
+		UpdateConfigSettings();
+		//SetStatus("ReceivedMaps", 0);
+	}
+}
+
+/**
+ * Method that's called to process a message that this plugin has
+ * subscribed for.  This particular method decodes the message and
+ * logs selective fields to a log file.
+ *
+ * @param msg Message that is received
+ * @routeable_message not used
+ */
+void MessageLoggerPlugin::HandleBasicSafetyMessage(BsmMessage &msg,
+		routeable_message &routeableMsg) {
+
+	char *BsmOut;
+	cJSON *BsmRoot, *BsmMessageContent, *_BsmMessageContent;
+
+	PLOG(logDEBUG)<<"HandleBasicSafetyMessage";
+	auto bsm = msg.get_j2735_data();
+
+	float speed_mph;
+	int32_t bsmTmpID;
+
+	bool isSuccess = false;
+	//asn_fprint(stdout, &asn_DEF_BasicSafetyMessage, bsm);
+	int32_t latitude = bsm->coreData.lat;
+	int32_t longitude = bsm->coreData.Long;
+	int32_t longAcceleration = bsm->coreData.accelSet.Long;
+
+	int32_t transtime = bsm->coreData.secMark;
+	uint16_t rawSpeed = bsm->coreData.speed;
+	uint16_t rawHeading = bsm->coreData.heading;
+	GetInt32((unsigned char *)bsm->coreData.id.buf, &bsmTmpID);
+
+	// Heading units are 0.0125 degrees.
+	float heading = rawHeading / 80.0;
+
+	// The speed is contained in bits 0-12.  Units are 0.02 meters/sec.
+	// A value of 8191 is used when the speed is not known.
+	if (rawSpeed != 8191)
+	{
+		// Convert from .02 meters/sec to mph.
+		speed_mph = rawSpeed / 50 * 2.2369362920544;
+
+		isSuccess = true;
+	}
+	else
+		speed_mph = 8191;
+
+	BsmRoot = cJSON_CreateObject(); // create root node
+	BsmMessageContent = cJSON_CreateArray(); // create root array
+
+	cJSON_AddItemToObject(BsmRoot, "BsmMessageContent", BsmMessageContent); // add BsmMessageContent array to Bsmroot
```

Comment:
This is new BSM JSON format for encoded BSM: 
If this MessageLoggerPlugin incorporate the real-time data format, he cJSON_AddItemObject should only include below fields:
`{ "BsmMessageContent": [{ "metadata": { "utctimestamp: "2020-11-30T23:45:24.913657Z" } "payload":"001480CF4B950C400022D2666E923D1EA6D4E28957BD55FFFFF001C758FD7E67D07F7FFF8000000002020218E1C1004A40196FBC042210115C030EF1408801021D4074CE7E1848101C5C0806E8E1A50101A84056EE8A1AB4102B840A9ADA21B9010259C08DEE1C1C560FFDDBFC070C0222210018BFCE309623120FFE9BFBB10C8238A0FFDC3F987114241610009BFB7113024780FFAC3F95F13A26800FED93FDD51202C5E0FE17BF9B31202FBAFFFEC87FC011650090019C70808440C83207873800000000001095084081C903447E31C12FC0"}]}`

Reply:
BSMMessageContent is an array, so the generated message content can also be like this:
`{ "BsmMessageContent": [{ "metadata": { "utctimestamp: "2020-11-30T23:45:24.913657Z" } "payload":"001480CF4B950C400022D2666E923D1EA6D4E28957BD55FFFFF001C758FD7E67D07F7FFF8000000002020218E1C1004A40196FBC042210115C030EF1408801021D4074CE7E1848101C5C0806E8E1A50101A84056EE8A1AB4102B840A9ADA21B9010259C08DEE1C1C560FFDDBFC070C0222210018BFCE309623120FFE9BFBB10C8238A0FFDC3F987114241610009BFB7113024780FFAC3F95F13A26800FED93FDD51202C5E0FE17BF9B31202FBAFFFEC87FC011650090019C70808440C83207873800000000001095084081C903447E31C12FC0"},
{ "metadata": { "utctimestamp: "2020-11-30T23:45:24.913657Z" } "payload":"001480CF4B950C400022D2666E923D1EA6D4E28957BD55FFFFF001C758FD7E67D07F7FFF8000000002020218E1C1004A40196FBC042210115C030EF1408801021D4074CE7E1848101C5C0806E8E1A50101A84056EE8A1AB4102B840A9ADA21B9010259C08DEE1C1C560FFDDBFC070C0222210018BFCE309623120FFE9BFBB10C8238A0FFDC3F987114241610009BFB7113024780FFAC3F95F13A26800FED93FDD51202C5E0FE17BF9B31202FBAFFFEC87FC011650090019C70808440C83207873800000000001095084081C903447E31C12FC0"}]}`

Reply:
in json utc is fine (human readable, but zulu time), for binary files unixtime is better